Serf cluster reconciler

bernardjkim opened the pull request:
## Description
This PR adds a serf cluster reconcile loop to planet. This loop periodically reconciles the serf cluster in case a serf member is ever partitioned off the cluster. The implementation relies on the kubernetes api to determine if the serf cluster is missing any members.

bernardjkim committed:
Add serf cluster reconciler

a-palchikov commented:
This should also attempt to remove members that aren't part of the cluster as per [this ticket](https://github.com/gravitational/gravity/issues/943) since we don't have serf encryption and cannot prevent partially cleaned up nodes from rejoining if they still run the planet container.

r0mant commented on tool/planet/agent.go:
```diff
@@ -364,6 +364,7 @@ func runAgent(conf *agent.Config, monitoringConf *monitoring.Config, leaderConf
 		}
 	}
 
+	go startSerfReconciler(ctx, &conf.SerfConfig)
```
I would modify the algorithm so it runs only on master nodes, otherwise we will be hammering API server in large clusters. We need to start keeping scaling in mind now.

r0mant commented on tool/planet/membership.go:
```diff
@@ -0,0 +1,157 @@
+// startSerfReconciler creates a control loop that periodically attempts to
+// reconcile serf cluster.
+func startSerfReconciler(ctx context.Context, serfConfig *serf.Config) {
+	ticker := time.NewTicker(30 * time.Second)
```
I think 30 seconds might be a bit too often for this type of check. I'd probably do it once every 5-10 minutes or so.

r0mant commented on tool/planet/membership.go:
```diff
@@ -0,0 +1,157 @@
+		return peers, trace.Wrap(err, "failed to get kubernetes clientset")
+	}
+	nodes, err := client.CoreV1().Nodes().List(metav1.ListOptions{
+		LabelSelector: advertiseIPKey,
```
What's this label selector for?

bernardjkim replied:
This is to ensure we are only selecting gravity nodes. Not sure if it is guaranteed that all nodes in the kubernetes cluster is part of the gravity cluster.

r0mant replied:
Yes, we do not support adding non-Gravity-managed nodes to the cluster, so I think this filter is redundant.

r0mant commented on tool/planet/membership.go:
```diff
@@ -0,0 +1,157 @@
+				log.WithError(err).Warn("Failed to reconcile serf cluster.")
+				break
+			}
+			log.WithField("num-joined", numJoined).Info("Reconciled serf cluster.")
```
I think logging the member that was missing and joined would be a bit more useful than just the number.

r0mant reviewed (commented):
> This should also attempt to remove members that aren't part of the cluster as per this ticket since we don't have serf encryption and cannot prevent partially cleaned up nodes from rejoining if they still run the planet container.

We need to be careful with this as this can potentially open a way to various race conditions where the agent e.g. may try to remove serf member while the new node is joining. I would keep it out of scope of what this PR tries to solve.

bernardjkim committed:
Add unit tests

bernardjkim committed:
Reconcile interval -> 10 min

bernardjkim committed:
Remove node label selector

bernardjkim committed:
Additional logs

- Log prev/new serf cluster members.
- Log if lingering serf node is detected.

r0mant reviewed (commented):
Have you tested how this behaves during shrink? To make sure it doesn't race with node removal and doesn't add the serf member back.

bernardjkim committed:
Read serf members before k8s nodes

- Gravity deletes nodes from k8s cluster before removing serf member.
This will help avoid race conditions when removing a node at the same
time as planet reconciles the serf cluster.

bernardjkim commented:
@r0mant 
As we discussed offline, the read order is now swapped and serf members is read before k8s nodes. 

We know that the k8s node is removed before the serf member is removed. So even though it is unlikely, if we encounter a race condition, at worst case, planet will just log that a serf member should no longer be a part of the cluster.

r0mant commented on tool/planet/membership.go:
```diff
@@ -0,0 +1,161 @@
+		return nil
+	}
+
+	if _, err := serfClient.Join(k8sPeers, false); err != nil {
```
I'm just curious, why are we joining to all peers, instead of having `shouldReconcileSerf` return a list of missing peers and join just those?

Then you wouldn't also need the `getSerfPeers` call below as you'd just be able to log the serf cluster and the missing peers.

bernardjkim replied:
Good point. No need to join all peers.

bernardjkim committed:
Join only missing peers

r0mant commented on tool/planet/membership.go:
```diff
@@ -0,0 +1,161 @@
+	for _, node := range nodes.Items {
+		addr, exists := node.Labels[advertiseIPKey]
+		if !exists {
+			return peers, trace.NotFound("node %s does not have %s label", node.Name, advertiseIPKey)
```
I would just log the warning and skip over the node if it doesn't have this label for some reason and return the rest. Also, how about failing over to `status.addresses` field in case this label doesn't exist? And skipping over the node if that's empty too.

bernardjkim committed:
Log warning if advertisedIP not specified
